File: cocmd/src/core/models/source_config_model.rs
use std::path::PathBuf;

use serde_derive::{Deserialize as De, Serialize as Se};
use tracing::error;

use super::script_model::{ScriptModel, StepModel};
use crate::utils::io::{from_file, from_yaml_file, normalize_path};
use crate::utils::sys::OS;

#[derive(Debug, Se, De, PartialEq, Eq, Hash, Clone)]
pub struct Automation {
    pub name: String,
    pub file: Option<String>,
    pub content: Option<ScriptModel>,
}

impl Automation {
    pub fn load_content(&self, location: &PathBuf) -> Automation {
        let mut automation_clone = self.clone();
        if let Some(file) = &self.file {
            let normalized_path = normalize_path(file, location);

            match from_yaml_file::<ScriptModel>(&normalized_path) {
                Ok(script_model) => {
                    let updated_steps: Vec<StepModel> = script_model
                        .steps
                        .iter()
                        .map(|step| {
                            if let Some(file) = &step.file {
                                let normalized_path = normalize_path(file, location);

                                match from_file(&normalized_path) {
                                    Ok(file_content) => {
                                        let mut step_clone = step.clone(); // Clone the existing StepModel
                                        step_clone.content = Some(file_content);
                                        step_clone
                                    }
                                    Err(err) => {
                                        // Handle the error if needed
                                        error!(err);
                                        step.clone() // Return the original step on error
                                    }
                                }
                            } else {
                                step.clone() // No file specified, return the original step
                            }
                        })
                        .collect();

                    // Update the content's steps with the updated_steps
                    let mut content_clone = script_model.clone();
                    content_clone.steps = updated_steps;
                    automation_clone.content = Some(content_clone);
                }
                Err(err) => {
                    // Handle the error if needed
                    error!(err);
                }
            }
        }
        automation_clone
    }

    pub fn supports_os(&self, os: &OS) -> bool {
        if let Some(content) = &self.content {
            if let Some(content_env) = &content.env {
                return *content_env == *os || *content_env == OS::ANY;
            }
        }
        false
    }
}

#[derive(Debug, Se, De, PartialEq, Eq, Hash)]
pub struct SourceConfigModel {
    pub name: String,
    pub aliases: Option<String>,
    pub paths: Option<Vec<String>>,
    pub automations: Option<Vec<Automation>>,
}

File: cocmd_old/cocmd_cli_py/core/models/source_config_model.py
from dataclasses import dataclass, field
from cocmd.core.os import OS
from cocmd.utils.io import DictLoader, YamlIO, normalize_path
from typing import List, Optional
from .script_model import ScriptModel


@dataclass
class Automation:
    name: str
    file: Optional[str]
    content: Optional[ScriptModel]

    def load_content(self, location: str):
        if self.file:
            self.content = YamlIO.from_file(
                normalize_path(self.file, location),
                cls=ScriptModel,
            )

    def supports_os(self, os: OS) -> bool:
        return self.content.env in (os, OS.ANY)


@dataclass
class SourceConfigModel(DictLoader):
    name: str
    aliases: Optional[str] = field(default_factory=str)
    paths: Optional[List[str]] = field(default_factory=list)
    automations: Optional[List[Automation]] = field(default_factory=list)
